### PR TITLE
[Dubbo-3106]Make getRegistered return unmodifiable collection. #3106

### PR DESCRIPTION
**Pull request by kezhenxu94**
## What is the purpose of the change

Related issue #3106 

## Brief changelog

XXXXX

## Verifying this change

XXXXX

Follow this checklist to help us incorporate your contribution quickly and easily:

- [x] Make sure there is a [GITHUB_issue](https://github.com/apache/incubator-dubbo/issues) field for the change (usually before you start working on it). Trivial changes like typos do not require a GITHUB issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [x] Format the pull request title like `[Dubbo-XXX] Fix UnknownException when host config not exist #XXX`. Each commit in the pull request should have a meaningful subject line and body.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/apache/incubator-dubbo/tree/master/dubbo-test).
- [x] Run `mvn clean install -DskipTests=false` & `mvn clean test-compile failsafe:integration-test` to make sure unit-test and integration-test pass.
- [ ] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/incubator-dubbo/wiki/Software-donation-guide).

### CODE REVIEW

**Comment by kezhenxu94** [comment]
There are nested collections in the return types of `org.apache.dubbo.registry.support.AbstractRegistry#getSubscribed` and `org.apache.dubbo.registry.support.AbstractRegistry#getNotified`, I'm wondering if it's worthwhile to wrap the nested collections too.

File: dubbo-registry/dubbo-registry-api/src/main/java/org/apache/dubbo/registry/support/AbstractRegistry.java [hunk 1]
```diff
@@ -119,15 +121,15 @@ protected void setUrl(URL url) {
     }
 
     public Set<URL> getRegistered() {
-        return registered;
+        return Collections.unmodifiableSet(registered);
```
Comment (khanimteyaz) [inline dubbo-registry/dubbo-registry-api/src/main/java/org/apache/dubbo/registry/support/AbstractRegistry.java]: @kezhenxu94  thanks for this PR. As you are returning the copy of collection, i think you can avoid making it unmodifiable.

Comment (satansk) [reply dubbo-registry/dubbo-registry-api/src/main/java/org/apache/dubbo/registry/support/AbstractRegistry.java]: Hi @khanimteyaz , it's just a wrapper, not a copy, so it's OK here :)

File: dubbo-registry/dubbo-registry-api/src/main/java/org/apache/dubbo/registry/support/AbstractRegistry.java [hunk 2]
```diff
@@ -50,6 +51,7 @@
 import java.util.concurrent.Executors;
 import java.util.concurrent.atomic.AtomicLong;
 import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
```
Comment (satansk) [inline dubbo-registry/dubbo-registry-api/src/main/java/org/apache/dubbo/registry/support/AbstractRegistry.java]: This line should be removed.

**Comment by satansk** [comment]
Overall, this looks good me, and @kezhenxu94 could you please fix the CI?

**Comment by kezhenxu94** [comment]
@satansk Will fix it soon

**Comment by kezhenxu94** [comment]
@satansk fixed

**Comment by kezhenxu94** [comment]
19 days passed, can any one of the admins verify this patch? @satansk or other admins

**Comment by CrazyHZM** [comment]
LGTM

**Pull request merged**